File: neetcode/blind75/intervals/non_overlapping_intervals.py
# https://neetcode.io/problems/non-overlapping-intervals

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        
        intervals.sort(key=lambda interval:interval[0])
        res = 0
        tailEnd = intervals[0][1]

        for start, end in intervals[1:]:
            if (start >= tailEnd):
                # no overlap, set new end int
                tailEnd = end
            else:
                # overlap.
                # removal count
                res += 1

                # new end should be the lowest int because it will have the least overlap potential
                tailEnd = min(tailEnd, end)
        
        return res

File: neetcode/neet150_JS/Intervals/non-overlapping-intervals.js
// https://neetcode.io/problems/non-overlapping-intervals

class Solution {
    /**
     * @param {number[][]} intervals
     * @return {number}
     */
    eraseOverlapIntervals(intervals) {
        if (intervals.length < 2) {
            return 0
        }

        intervals.sort((a, b) => {return a[0] - b[0]})
        let removeCnt = 0
        let prevEnd = intervals[0][1]

        for (let i = 1; i < intervals.length; i ++) {
            const start = intervals[i][0]
            const end = intervals[i][1]

            if (start >= prevEnd) {
                prevEnd = end
            } else {
                removeCnt += 1
                prevEnd = Math.min(prevEnd, end)
            }
        }

        return removeCnt
    }
}
